src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java

```diff
@@ -0,0 +1,161 @@
+                    topDoc.doc = doc + docBase;
+                    topDoc.score = score;
+                    pq.updateTop();
+                }
+            }
+        };
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return hitsThresholdChecker.scoreMode();
+    }
+
+    protected void updateMinCompetitiveScore(Scorable scorer) throws IOException {
+        if (hitsThresholdChecker.isThresholdReached() && pqTop != null && pqTop.score != Float.NEGATIVE_INFINITY) { // -Infinity is the
+                                                                                                                    // boundary score
+            // we have multiple identical doc id and collect in doc id order, we need next float
+            float localMinScore = Math.nextUp(pqTop.score);
+            if (localMinScore > minCompetitiveScore) {
+                scorer.setMinCompetitiveScore(localMinScore);
+                totalHitsRelation = TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO;
+                minCompetitiveScore = localMinScore;
+            }
+        }
+    }
+
+    /**
+     * Get resulting collection of TopDocs for hybrid query after we ran search for each of its sub query
+     * @return
+     */
+    public TopDocs[] topDocs() {
+        TopDocs[] topDocs = new TopDocs[compoundScores.length];
+        for (int i = 0; i < compoundScores.length; i++) {
+            int qTopSize = totalHits[i];
+            TopDocs topDocsPerQuery = topDocsPerQuery(0, Math.min(qTopSize, compoundScores[i].size()), compoundScores[i], qTopSize);
+            topDocs[i] = topDocsPerQuery;
+        }
+        return topDocs;
+    }
+
+    private TopDocs topDocsPerQuery(int start, int howMany, PriorityQueue<ScoreDoc> pq, int totalHits) {
+        int size = howMany;
+
+        if (howMany < 0) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "Number of hits requested must be greater than 0 but value was %d", howMany)
+            );
+        }
+
+        if (start < 0) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "Expected value of starting position is between 0 and %d, got %d", size, start)
+            );
+        }
+
+        if (start >= size || howMany == 0) {
+            return EMPTY_TOPDOCS;
+        }
+
+        howMany = Math.min(size - start, howMany);
+        ScoreDoc[] results = new ScoreDoc[howMany];
+        // pq's pop() returns the 'least' element in the queue, therefore need
+        // to discard the first ones, until we reach the requested range.
+        for (int i = pq.size() - start - howMany; i > 0; i--) {
+            pq.pop();
+        }
+
+        // Get the requested results from pq.
+        populateResults(results, howMany, pq);
+
+        return new TopDocs(new TotalHits(totalHits, totalHitsRelation), results);
+    }
+
+    protected void populateResults(ScoreDoc[] results, int howMany, PriorityQueue<ScoreDoc> pq) {
```

Comment:
No need to check if `results` size is big enough for howMany value and `pq` also has enough items in it?

Reply:
it's checked in previous logic, but let me create one more explicit check here